tests/support/mod.rs

```diff
@@ -19,40 +19,52 @@ use std::str;
 use std::sync::{Arc, Condvar, Mutex};
 use std::thread;
 use std::time::Duration;
-
+use std::panic;
 
 use support::paths::TestPathExt;
 
 pub mod paths;
 
 /// Executes `func` and panics if it takes longer than `dur`.
 pub fn timeout<F>(dur: Duration, func: F)
-    where F: FnOnce() + Send + 'static {
-    let pair = Arc::new((Mutex::new(false), Condvar::new()));
+    where F: FnOnce() + Send + 'static + panic::UnwindSafe {
+    let pair = Arc::new((Mutex::new(TestState::Running), Condvar::new()));
     let pair2 = pair.clone();
 
     thread::spawn(move|| {
         let &(ref lock, ref cvar) = &*pair2;
-        func();
-        let mut finished = lock.lock().unwrap();
-        *finished = true;
+        match panic::catch_unwind(|| func()) {
+            Ok(_) => *lock.lock().unwrap() = TestState::Success,
+            Err(_) => *lock.lock().unwrap() = TestState::Fail,
```

Comment:
Could we propagate the cause to TestState::Fail and later to https://github.com/rust-lang-nursery/rls/pull/668/files#diff-c2fa527c5defdbe0ec9a96f9aa02b5d9R55?

Reply:
There isn't any info in the [result type](https://doc.rust-lang.org/std/thread/type.Result.html), it just tells us the function panicked. Panicking again is the best we can do, failing the test and drawing attention to the backtrace of the cause.

Reply:
Ah, right, fair enough!